### PR TITLE
Support loading Projectors

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

Depends on 
* https://github.com/gazebosim/gz-rendering/pull/845
* https://github.com/gazebosim/sdformat/pull/1277

## Summary

Added support for loading Projectors.
* Updated SdfEntityCreator to create projector entities
* Updated rendering component to load and create projector objects in the scene
* Updated scene broadcaster to publish scene info projectors
* Added conversion function for serializing / deserializing `sdf::Projector` to / from `msgs::Projector`

## Test it

Added and updated various unit and integration tests

Run the demo world:

```
gz sim -v 4 projector.sdf
```

This should go and download the [Camera with Projector](https://app.gazebosim.org/OpenRobotics/fuel/models/Camera%20with%20Projector) from fuel. The model consists of 2 IR cameras, 1 IR pattern projector, and 1 RGB camera. The IR pattern should only be visible to the 2 IR cameras and invisible to the RGB camera.

projector_world

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/gz/sim/components/Projector.hh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#ifndef GZ_SIM_COMPONENTS_PROJECTOR_HH_
+#define GZ_SIM_COMPONENTS_PROJECTOR_HH_
+
+#include <gz/msgs/projector.pb.h>
+
+#include <sdf/Projector.hh>
+
+#include <gz/sim/components/Factory.hh>
+#include <gz/sim/components/Component.hh>
+#include <gz/sim/components/Serialization.hh>
```
Comment (ahcorde) [inline include/gz/sim/components/Projector.hh:26]: alphabetize

File: include/gz/sim/rendering/SceneManager.hh [hunk 2]
```diff
@@ -269,6 +269,14 @@ inline namespace GZ_SIM_VERSION_NAMESPACE {
     public: rendering::ParticleEmitterPtr UpdateParticleEmitter(Entity _id,
         const msgs::ParticleEmitter &_emitter);
 
+    /// \brief Create a projector
+    /// \param[in] _id Unique projector id
+    /// \param[in] _emitter Projector sdf dom
```
Comment (ahcorde) [inline include/gz/sim/rendering/SceneManager.hh:274]:
```suggestion
    /// \param[in] _projector Projector sdf dom
```

File: src/SdfEntityCreator.cc [hunk 3]
```diff
@@ -73,6 +73,7 @@
 #include "gz/sim/components/ParentLinkName.hh"
 #include "gz/sim/components/ParentEntity.hh"
 #include <gz/sim/components/ParticleEmitter.hh>
+#include <gz/sim/components/Projector.hh>
```
Comment (ahcorde) [inline src/SdfEntityCreator.cc:76]: alphabetize

File: src/rendering/RenderUtil.cc [hunk 4]
```diff
@@ -1274,6 +1282,14 @@ void RenderUtil::Update()
...
 
+    for (const auto &projector : newProjectors)
+    {
+      this->dataPtr->sceneManager.CreateProjector(
+          std::get<0>(projector), std::get<1>(projector),
+          std::get<2>(projector));
+    }
+
+
     if (this->dataPtr->enableSensors && this->dataPtr->createSensorCb)
```
Comment (ahcorde) [inline src/rendering/RenderUtil.cc:1293]:
```suggestion
    }

    if (this->dataPtr->enableSensors && this->dataPtr->createSensorCb)
```

File: src/rendering/SceneManager.cc [hunk 5]
```diff
@@ -1657,6 +1663,65 @@ rendering::ParticleEmitterPtr SceneManager::UpdateParticleEmitter(Entity _id,
...
+    Entity _id, const sdf::Projector &_projector, Entity _parentId)
+{
+  if (!this->dataPtr->scene)
+    return rendering::ProjectorPtr();
+
+  if (this->dataPtr->projectors.find(_id) !=
+     this->dataPtr->projectors.end())
+  {
+    gzerr << "Projector with Id: [" << _id << "] already exists in the "
+           <<" scene" << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:1677]:
```suggestion
    gzerr << "Projector with Id: [" << _id << "] already exists in the "
          << "scene" << std::endl;
```

File: test/integration/projector.cc [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+/////////////////////////////////////////////////
+// Load an SDF with a projector and verify its properties.
+// See https://github.com/gazebosim/gz-sim/issues/1175
+TEST_F(ProjectorTest, GZ_UTILS_TEST_DISABLED_ON_WIN32(SDFLoad))
+{
+  bool projectorChecked{false};
+
+  this->LoadWorld("test/worlds/projector.sdf");
```
Comment (ahcorde) [inline test/integration/projector.cc:67]: `common::joinPaths`

File: test/integration/projector.cc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+                EXPECT_EQ("projector", _name->Data());
+                EXPECT_EQ(_name->Data(), _projector->Data().Name());
+                EXPECT_EQ(math::Pose3d(0, 1, 0, 0, 0, 0), _pose->Data());
+                EXPECT_EQ(_pose->Data(), _projector->Data().RawPose());
+                EXPECT_DOUBLE_EQ(2.0, _projector->Data().NearClip());
+                EXPECT_DOUBLE_EQ(7.0, _projector->Data().FarClip());
+                EXPECT_EQ(math::Angle(0.5), _projector->Data().HorizontalFov());
+                EXPECT_EQ(0x01, _projector->Data().VisibilityFlags());
+                EXPECT_EQ("path/to/dummy_image.png",
```
Comment (ahcorde) [inline test/integration/projector.cc:94]: `common::joinPaths`

File: test/integration/scene_broadcaster_system.cc [hunk 8]
```diff
@@ -1030,6 +1030,66 @@ TEST_P(SceneBroadcasterTest,
   EXPECT_EQ(1, count);
 }
 
+TEST_P(SceneBroadcasterTest,
+    GZ_UTILS_TEST_DISABLED_ON_WIN32(SceneInfoHasProjector))
+{
+  // Start server
+  gz::sim::ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      common::joinPaths("/", "test", "worlds", "projector.sdf"));
```
Comment (ahcorde) [inline test/integration/scene_broadcaster_system.cc:1039]:
```suggestion
      common::joinPaths("test", "worlds", "projector.sdf"));
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: test/integration/scene_broadcaster_system.cc (hunk 8 above)
Comment (iche033) [reply test/integration/scene_broadcaster_system.cc]: 12cf0015552ea24935963306eea3c7f231395bea

File: test/integration/projector.cc (hunk 7 above)
Comment (iche033) [reply test/integration/projector.cc]: 12cf0015552ea24935963306eea3c7f231395bea

File: test/integration/projector.cc (hunk 6 above)
Comment (iche033) [reply test/integration/projector.cc]: 12cf0015552ea24935963306eea3c7f231395bea

File: src/rendering/SceneManager.cc (hunk 5 above)
Comment (iche033) [reply src/rendering/SceneManager.cc]: 12cf0015552ea24935963306eea3c7f231395bea

File: src/rendering/RenderUtil.cc (hunk 4 above)
Comment (iche033) [reply src/rendering/RenderUtil.cc]: 12cf0015552ea24935963306eea3c7f231395bea

File: include/gz/sim/rendering/SceneManager.hh (hunk 2 above)
Comment (iche033) [reply include/gz/sim/rendering/SceneManager.hh]: 12cf0015552ea24935963306eea3c7f231395bea

File: include/gz/sim/components/Projector.hh (hunk 1 above)
Comment (iche033) [reply include/gz/sim/components/Projector.hh]: 12cf0015552ea24935963306eea3c7f231395bea

**Comment by iche033** [comment]
still needs sdf13 release

File: src/Conversions.cc [hunk 9]
```diff
@@ -1752,6 +1753,63 @@ sdf::ParticleEmitter gz::sim::convert(const msgs::ParticleEmitter &_in)
...
+      try
+      {
+        out.SetVisibilityFlags(std::stoul(data.value(0)));
+      }
+      catch (...)
+      {
+        gzerr << "Failed to parse projector <visibility_flags>: "
+              <<  data.value(0) << ". Using default value: 0xFFFFFFFF."
+              << std::endl;
+        out.SetVisibilityFlags(0xFFFFFFFF);
```
Comment (ahcorde) [inline src/Conversions.cc:1805]:
```suggestion
        gzerr << "Failed to parse projector <visibility_flags>: "
              << data.value(0) << ". Using default value: 0xFFFFFFFF."
              << std::endl;
        out.SetVisibilityFlags(0xFFFFFFFF);
```

File: src/systems/scene_broadcaster/SceneBroadcaster.cc [hunk 10]
```diff
@@ -1079,6 +1089,29 @@ void SceneBroadcasterPrivate::SceneGraphAddEntities(
...
+    components::Pose>(
+      [&](const Entity &_entity,
+          const components::Projector *_projectorComp,
+          const components::ParentEntity *_parentComp,
+          const components::Pose *_poseComp) -> bool
+      {
+        auto projectorMsg = std::make_shared<msgs::Projector>();
+        projectorMsg->CopyFrom(
+            convert<msgs::Projector>(_projectorComp->Data()));
+        // \todo(anyone) add id field to projector msg
```
Comment (ahcorde) [inline src/systems/scene_broadcaster/SceneBroadcaster.cc:1103]: Add issue in gz-msgs ?

Comment (iche033) [reply src/systems/scene_broadcaster/SceneBroadcaster.cc:1103]: yep, ticketed issue https://github.com/gazebosim/gz-msgs/issues/341

File: src/Conversions.cc (hunk 9 above)
Comment (iche033) [reply src/Conversions.cc]: [3c4e8c8](https://github.com/gazebosim/gz-sim/pull/1979/commits/3c4e8c87f47c4374441b1f79a78a4fc0eb5e5ba6)

**Pull request merged**